cartex/src/runtime/mpi_comm/runtime.cpp

```diff
@@ -114,34 +138,331 @@ runtime::impl::neighborhood::neighborhood(
     }
 }
 
+int
+runtime::impl::neighborhood::sendtag(int field_id, int dim, bool left) const noexcept
+{
+    return field_id + num_fields * (dim * 2 + (left ? 0 : 1)) + num_fields * 6 * d.thread;
+}
+
+int
+runtime::impl::neighborhood::recvtag(int field_id, int dim, bool left) const noexcept
+{
+    const auto tid_l = (dim == 0 ? x_r.thread : (dim == 1 ? y_r.thread : z_r.thread));
+    const auto tid_r = (dim == 0 ? x_l.thread : (dim == 1 ? y_l.thread : z_l.thread));
+    return field_id + num_fields * (dim * 2 + (left ? 0 : 1)) +
+           num_fields * 6 * (left ? tid_l : tid_r);
+}
+
+void
+runtime::impl::neighborhood::exchange(memory_type& field, int field_id)
+{
+    runtime::real_type* data = field.hd_data();
+
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *x_send_l, x_l.rank, sendtag(field_id, 0, true),
+        data, 1, *x_recv_r, x_r.rank, recvtag(field_id, 0, true), comm, MPI_STATUS_IGNORE));
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *x_send_r, x_r.rank, sendtag(field_id, 0, false),
+        data, 1, *x_recv_l, x_l.rank, recvtag(field_id, 0, false), comm, MPI_STATUS_IGNORE));
+
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *y_send_l, y_l.rank, sendtag(field_id, 1, true),
+        data, 1, *y_recv_r, y_r.rank, recvtag(field_id, 1, true), comm, MPI_STATUS_IGNORE));
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *y_send_r, y_r.rank, sendtag(field_id, 1, false),
+        data, 1, *y_recv_l, y_l.rank, recvtag(field_id, 1, false), comm, MPI_STATUS_IGNORE));
+
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *z_send_l, z_l.rank, sendtag(field_id, 2, true),
+        data, 1, *z_recv_r, z_r.rank, recvtag(field_id, 2, true), comm, MPI_STATUS_IGNORE));
+    CARTEX_CHECK_MPI_RESULT(MPI_Sendrecv(data, 1, *z_send_r, z_r.rank, sendtag(field_id, 2, false),
+        data, 1, *z_recv_l, z_l.rank, recvtag(field_id, 2, false), comm, MPI_STATUS_IGNORE));
+}
+
+#ifdef __CUDACC__
+__global__ void
+pack_x_kernel(runtime::real_type const* field, runtime::real_type* buffer_left,
```

Comment:
This structure is fine of course, but I would suggest we try another version, in which we iterate over the Z-dimension inside the kernel. Not sure it will be faster, but if it's not too much work, I think it would be worth a try. In my prior tests with stencil computations (using the whole cube) that was the preferred approach. Just halo exchange is of course different, but maybe for large-enough cubes we might see some improvement due to kernel startup savings.

Reply:
I can try this. Note, that z direction is the dimension with largest stride - so each gpu thread would jump around in memory a lot. Is that what you are suggesting?

Reply:
yes. jumping in memory doesn't matter: you anyway have to traverse Z, you just do it by spawning thread blocks. The memory still has to be accessed.

But again, there is much less parallelism for halo exchange than for whole stencil computations, so I'm not sure this will be beneficial in any way. You will save on kernel startups, but you will spawn much fewer thread blocks.

Reply:
I don't think I will save on startup - it is 1 kernel after all. but I can try and see.